### DISCUSSION TITLE
Reuse existing event loop if one is already running

### DISCUSSION THREAD

**Discussion by medihack** [pytest-dev/pytest-asyncio#546] [Q&A] [answered]
When I run async tests using `pytest-asyncio` after integration tests using `playwright-pytest` (only using playwright.sync_api calls), then those tests fail with the error already described in #359 . I don't know much about the internals of `pytest-asyncio`, but wouldn't it be possible to reuse the event loop when one is already running (and only try to start a new one when not)?

**Comment by medihack** [comment]
Just as a note (to help others that stumble upon this) there is a simple workaround. Install [nest_asyncio](https://pypi.org/project/nest-asyncio/) and then put this into the root `conftest.py`:

```
def pytest_configure():
    nest_asyncio.apply()
```

**Comment by seifertm** [comment]
If I understand correctly, your question is why pytest-asyncio cannot reuse an already running loop. That's a fair point to make.

When writing unit tests, it can be a good idea to test a unit in isolation. In order to exclude any potential side-effects, pytest-asyncio tries to run the test in a fresh event loop.

By design, the event loops in asyncio are limited to one per thread. The loop schedules all the coroutines and other awaitables in a thread. In order to do so, it monitors things like network sockets and forwards data to the corresponding coroutine whenever new data is available. This is why loops cannot be easily nested. There are some ways around it, e.g. nest_asyncio or aioloop-proxy, but none of them are currently integrated into pytest-asyncio,

You can reuse an event loop over multiple tests by reimplementing the `event_loop` fixture provided by pytest-asyncio with a larger scop, e.g. "module" to reuse it over all tests of a module. Please refer to the docs https://pytest-asyncio.readthedocs.io/en/latest/reference/fixtures.html#event-loop for an example.

Does this explanation help?

**Comment by martinkirch** [comment, marked as answer]
Hello,
I've encountered the same case, it was actually caused by `playwright-pytest` as exposed in https://github.com/microsoft/playwright-pytest/issues/240 (this discussion was also submitted as https://github.com/microsoft/playwright-pytest/issues/167). On the pytest-asyncio side, reimplementing the `event_loop` fixture is deprecated since #587 .

Now we have another solution: switching to [pytest-playwright-asyncio](https://pypi.org/project/pytest-playwright-asyncio/). That implies updating existing tests to import from `playwright.async_api`, which also requires adding _many_ `async/await` and  [putting all tests in the same loop](https://pytest-asyncio.readthedocs.io/en/latest/how-to-guides/run_session_tests_in_same_loop.html) (at least because Playwright's `Browser` is a session-wide fixture that tries to access the event loop... but setting `asyncio_default_fixture_loop_scope = "session"` was not enough). Its documentation is expected [on this page](https://playwright.dev/python/docs/test-runners) in a near future.

**Reply by seifertm** [reply]
Thanks for pointing this out @martinkirch !
It's nice to see that pytest-playwright-asyncio was able to use pytest-asyncio under the hood.

**Comment by LouisJULIEN** [comment]
I fixed it this way https://github.com/pytest-dev/pytest-asyncio/issues/947#issuecomment-2525072042

```python
@pytest.fixture(scope="module")
def event_loop():
    loop = asyncio.get_event_loop()
    yield loop
    loop.close()
```
